File: 1_rust_safety/c_examples/escaped_reference.c
#include <stdlib.h>
#include <stdio.h>

struct S {
    int a;
    float b;
};


void foo(struct S* s){
  printf("%d\n",s -> a);
}

struct S okay() {
   struct S bar = (struct S){1, 3.0};
   foo(&bar);
   return bar;
}

struct S* also_okay() {
    struct S* s = malloc(sizeof(struct S));
    s->a = 1;
    s->b = 3.0;
    return s;
}

struct S* bad () {
   struct S foo = (struct S){1, 3.0};
   return &foo;
}

int main() {
}

File: 1_rust_safety/src/escaped_reference.rs
struct S {
    a: i32,
    b: f32,
}

fn foo(s: &S) {
    println!("{}", s.a);
}

fn okay() -> S {
    S { a: 1, b: 3.0 }
}

fn also_okay() -> Box<S> {
    Box::new(S { a: 1, b: 3.0 })
}

fn bad() -> &S {
    &S { a: 1, b: 3.0 }
}